fontforge/splinestroke.c

```diff
@@ -32,3689 +32,1805 @@
+enum ShapeType NibIsValid(SplineSet *ss) {
+    Spline *s;
+    BasePoint lref_cp;
+    bigreal d, anglesum = 0, angle, last_angle;
+    int n = 1;
 
-	off = (c->corners[i].x*slope->y) - (c->corners[i].y*slope->x);
-	if ( off>0 && off>=bestr_off ) {
-	    if ( off>bestr_off ) {
-		bestr_off = off;
-		bestr = i;
-		bestr2 = -1;
-	    } else {
-		bestr2 = i;
-	    }
-	}
-    }
-    if ( bestl==-1 || bestr==-1 )
-	IError( "Failed to find corner in WhichPolyCorner" );
-    if ( bestl2!=-1 ) {
-	if ( bestl+1!=bestl2 && !(bestl2==c->n-1 && bestl==0) )
-	    IError("Unexpected multiple left corners in WhichPolyCorner");
-	if ( bestl==0 && bestl2==c->n-1 )
-	    bestl = c->n-1;
+    if ( ss->first==NULL )
+	return Shape_TooFewPoints;
+    if ( ss->first->prev==NULL )
+	return Shape_NotClosed;
+    if ( !SplinePointListIsClockwise(ss) )
+	return Shape_CCW;
+
+    s = ss->first->prev;
+    last_angle = atan2(s->to->me.y - s->from->me.y,
+                       s->to->me.x - s->from->me.x);
+    if ( SplineIsLinear(s) )
+	lref_cp = s->from->me;
+    else
+	lref_cp = s->to->prevcp;
+
+    s = ss->first->next;
+    SplinePointListSelect(ss, false);
+    SplinePointListClearCPSel(ss);
+    // Polygonal checks
+    while ( true ) {
+	s->from->selected = true;
+	if ( BPWITHIN(s->from->me, s->to->me,1e-2) )
+	    return Shape_TinySpline;
+	angle = atan2(s->to->me.y - s->from->me.y, s->to->me.x - s->from->me.x);
+	if ( RealWithin(angle, last_angle, 1e-4) )
+	    return Shape_PointOnEdge;
+	d = last_angle-angle;
+	d = NORMANGLE(d);
+	if ( d<0 )
+	    return Shape_CCWTurn;
+	anglesum += d;
+	s->from->selected = false;
+	last_angle = angle;
+	s=s->to->next;
+	if ( s==ss->first->next )
+	    break;
+	++n;
     }
-    if ( bestr2!=-1 ) {
-	if ( bestr+1!=bestr2 && !(bestr2==c->n-1 && bestr==0) )
-	    IError("Unexpected multiple right corners in WhichPolyCorner");
-	if ( bestr==0 && bestr2==c->n-1 )
-	    bestr = c->n-1;
+    if ( n<3 )
+	return Shape_TooFewPoints;
+    if ( !RealWithin(anglesum, 2*FF_PI, 1e-1) )
+	return Shape_SelfIntersects;
+
+    assert( s==ss->first->next );
+    // Curve/control point checks
+    while ( true ) {
+	if ( SplineIsLinear(s) )
+	    lref_cp = s->from->me;
+	else {
+	    if ( s->from->nonextcp || BPNEAR(s->from->nextcp, s->from->me) ||
+	         s->to->noprevcp || BPNEAR(s->to->prevcp, s->to->me) )
+		return Shape_HalfLinear;
+	    s->from->nextcpselected = true;
+	    if ( LineSameSide(s->from->me, s->to->me, s->from->nextcp,
+	                      s->from->prev->from->me, true) )
+		return Shape_BadCP;
+	    if ( !LineSameSide(lref_cp, s->from->me,
+	                       s->from->nextcp, s->to->me, true) )
+		return Shape_BadCP;
+	    if ( !LineSameSide(s->to->me, s->to->prevcp,
+	                       s->from->nextcp, s->from->me, true) )
+		return Shape_BadCP;
+	    s->from->nextcpselected = false;
+	    s->from->selected = false;
+	    s->to->selected = true;
+	    s->to->prevcpselected = true;
+	    if ( LineSameSide(s->fro
```

Comment:
Is this for debugging purposes?

Reply:
Not exactly. 

Cusps never show up in the finished product -- they're always "under the nib" -- so there's no real reason to trace them accurately. However, I started out doing that anyway because it didn't affect the performance much and I figured someone might want to know the exact-ish cusp geometry. 

Unfortunately there's some weird issue with the fitting algorithm that always co-locates one control point at the start, eliminating the slope. This makes the curve both inaccurate and tends to confuse Remove Overlap. 

So I switched the code to just put a few points on the cusps but left the means there so that if someone wanted to go back and look at the tracing problem it was easy.